### ISSUE TITLE
How should a client deal with links and operations based on the current spec?

### ISSUE THREAD

**Issue by lambdakris** [HydraCG/gitbook#21]
I have been going through the diff resources available to get a better sense of how a client would use hypermedia and since most of the practical examples are here in the cookbook and the heracles.ts integration tests, I'm finding alot of usage of "links" and "collections" which I'm wondering how those would translate to the current version of the draft and how a more current client would perform discovery without explicit links and collections?

**Comment by alien-mcl** [comment]
There are two places where a client can find hypermedia:
- API documentation
- inline

The former is pretty straight forward - an API documents ad-hoc whatever it can/wants to, i.e. supported classes, their operations, entry points etc.

The latter is a mixture of a requested resource, optional additional resources and hypermedia describing how requested resource (and possibly other resources attached) are related to other resources (links) and operations. The aforementioned `hydra:collection` is a specialized link that implicitly tells the client that the resource linked returns a `hydra:Collection` of some sort. Other links are not so descriptive. 

As for the operations, the `hydra:operation` tells the client that the resource on the left of that relation has an operation available.

Whole discovering is not a simple task - that's why Heracles.ts exists so your client could be built on top of it without need to implement that logic.

**Comment by tpluscode** [comment]
What do you mean by "without explicit links and collections"?

tl;dr; given a dereferenced resource, you would do two things:

1. look for properties which link to other resources. `hydra:search` and `hydra:collection` are such built-in relations
2. dereference the API Documentation to find out what operations are possible for any give resource

You might also have a look at [alcaeus](https://npm.im/alcaeus) (I'm the author). It has documentation and live examples on https://alcaeus.hydra.how.

**Comment by lambdakris** [comment]
Thank you both for replies! I think I have a better idea of what I should be doing now. I think my confusion was around instances vs classes, where some of the examples in the cookbook were using "collections" and "links" as terms, and so I figured that was what the client relied on for parsing/discovery, but was not sure how this translated to the newer examples which did not use "collections" and "links" as terms while the client still exposed fields for collections and links. But based on your answers I think I understand that the client uses the context to determine what goes into the "collections" and "links" fields.
